### ISSUE TITLE
Questions to flash optee os into hikey lemaker version board

### ISSUE THREAD

**Issue by gssrok** [OP-TEE/optee_os#815]
Hi.
I've tried to flash optee os to my hikey board(lemaker ver., 2G RAM, 8G flash, debian).

But when I compiled source codes in my host pc(ubuntu 12.04.5 LTS) and flash binaries into the board, I faced some suspicious messages.

1. When I put make all after make system-img, error is reported.

```
testRandomDataCycles (TianoCompress.Tests) ... Traceback (most recent call last):
  File "/home/gsseo/devel/optee/build/../optee_os/out/arm/export-ta_arm32/scripts/sign.py", line 80, in <module>
    main()
  File "/home/gsseo/devel/optee/build/../optee_os/out/arm/export-ta_arm32/scripts/sign.py", line 40, in main
    from Crypto.Signature import PKCS1_v1_5
ImportError: No module named Signature
make[3]: *** [/home/gsseo/devel/optee/build/../optee_test/out/ta/create_fail_test/c3f6e2c0-3548-11e1-b86c0800200c9a66.ta] Error 1
make[3]: Leaving directory `/home/gsseo/devel/optee/optee_test/ta/create_fail_test'
make[2]: *** [ta] Error 2
make[2]: Leaving directory `/home/gsseo/devel/optee/optee_test/ta'
make[1]: *** [ta] Error 2
make[1]: Leaving directory `/home/gsseo/devel/optee/optee_test'
make: *** [xtest-common] Error 2
make: *** Waiting for unfinished jobs....
ok
```

Q. Can those errors be ignored?

2. After make all, I ignored errors reported and put make flash. And then following messages are printed.

```
Jumper 1-2: Closed (Auto power up = Boot up when power is applied)
       3-4: Open   (Boot Select = Boot from eMMC)
       5-6: Closed (GPIO3-1 = Low: UEFI runs Fastboot app)
1. Connect USB OTG cable, the micro USB cable (press any key)
2. Connect HiKey to power up (press any key)
3. Wait until you see the (UART) message
    "Android Fastboot mode - version x.x Press any key to quit."
   Then press any key to continue flashing
fastboot flash ptable /home/gsseo/devel/optee/build/../l-loader/ptable-linux-4g.img
< waiting for device >
```

Q. As illustrated in the 96 boards wiki, jumper 1-2 and 3-4 should be closed to flash binaries into the board. However make file prints above messages. What pins should I close?

Q. My board has 8G flash. But make file requests partition table image for 4G board. Also, In the l-loader directory, there's no image file! What shall I do?

Thank you.

**Comment by jforissier** [comment]
Regarding the first error: your python is missing the Crypto.Signature. Maybe because the Crypto package is no installed, or maybe because it is too old and does not contain Signature.
Try `sudo apt-get python-crypto`, and if it doesn't help, you may try `pip install --upgrade --user pycrypto` to install a local copy.
But, I would recommend you upgrade your Ubuntu distribution if you can ;)

**Comment by vchong** [comment]
Q2: There are 2 'modes' you can use to flash binaries into the mode. Jumper 1-2 and 3-4 closed put you into 'recovery' mode, corresponding to `make recovery` in the makefile. 1-2 and 5-6 closed put you into 'fastboot' mode, corresponding to `make flash`, but it requires an UEFI image (fip.bin) that supports 'fastboot' mode already flashed into the device prior. If you're not sure, go with 'recovery' mode.

Q3. See below for guide to adjust the makefile for 8g flash.
https://github.com/OP-TEE/build/issues/64#issuecomment-219764614

**Comment by gssrok** [comment]
Thank you for your kind answers!
I'd try it again with those answers.

**Comment by gssrok** [comment]
I updated my ubuntu to 16.04 LTS and tried again to flash to my hikey board.
However an another problem comes.

While I put 'make all' to build EDK2 and other binaries, this command reported building EDK2 is failed.

```
"GenFw" -e SEC -o $HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/ArmPlatformPkg/PrePi/PeiMPCore/DEBUG/ArmPlatformPrePiMPCore.efi $HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/ArmPlatformPkg/PrePi/PeiMPCore/DEBUG/ArmPlatformPrePiMPCore.dll
GenFw: ERROR 3000: Invalid
  WriteSections64(): $HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/ArmPlatformPkg/PrePi/PeiMPCore/DEBUG/ArmPlatformPrePiMPCore.dll unsupported ELF EM_AARCH64 relocation R_AARCH64_ADR_PREL_PG_HI21.
GenFw: ERROR 3000: Invalid
  WriteSections64(): $HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/ArmPlatformPkg/PrePi/PeiMPCore/DEBUG/ArmPlatformPrePiMPCore.dll unsupported ELF EM_AARCH64 relocation 0x11e.

....

GNUmakefile:397: recipe for target '$HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/ArmPlatformPkg/PrePi/PeiMPCore/DEBUG/ArmPlatformPrePiMPCore.efi' failed
make[2]: Leaving directory '$HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/ArmPlatformPkg/PrePi/PeiMPCore'


build.py...
 : error 7000: Failed to execute command
	make tbuild [$HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/ArmPlatformPkg/PrePi/PeiMPCore]


build.py...
 : error 7000: Failed to execute command
	make tbuild [$HOME/devel/optee/edk2/Build/HiKey/RELEASE_GCC49/AARCH64/MdeModulePkg/Core/Dxe/DxeMain]


build.py...
 : error F002: Failed to build module
	$HOME/devel/optee/edk2/ArmPlatformPkg/PrePi/PeiMPCore.inf [AARCH64, GCC49, RELEASE]

- Failed -
Build end time: 17:28:23, May.30 2016
Build total time: 00:00:31

HisiPkg/HiKeyPkg/Makefile:61: recipe for target 'all' failed
make[1]: *** [all] Error 1
make[1]: Leaving directory '$HOME/devel/optee/edk2'
common.mk:171: recipe for target 'edk2-common' failed
make: *** [edk2-common] Error 2
```


I didn't build EDK2 before this time.
Shall I build it according to [here](https://github.com/96boards/documentation/wiki/HiKeyUEFI)?

**Comment by gssrok** [comment]
I resolved EDK2 cross compile problem by replacing {OPTEE}/build/Makefile define edk2-call part's AARCH64_CROSS_COMPILE to LEGACY_AARCH64_CROSS_COMPILE

In the toolchain.mk file, they said legacy compiler (linaro gcc 4.9) should be used due to relocation error about edk2. So I can build EDK2.

Also, there is an another problem about openssl header file.
So I installed libssl-dev.

At this point, I finished make all command and rewrite hikey_debian.mk's make flash section like that adding sudo before each fastboot command.

I guess I would close this issue however if I face some problems later, will reopen this. Thank you.
